### ISSUE TITLE
Configuring incomplete, errors occurred

### ISSUE THREAD

**Issue by schkovich** [arobenko/cc.mqtt5.libs#2]
Hello, there!

OS: Ubuntu 24.04 server.

CMAKE: 3.28.3

I cloned the repo and ran `BUILD_DIR=/home/ubuntu/cc.mqtt5.libs ./prepare_externals.sh`.

I created the build directory and ran from it:
```sh
export PROOT="/home/ubuntu/cc.mqtt5.libs"
export EXT_DIR="${PROOT}/externals"
export COMMS="${EXT_DIR}/comms/install"
export MQTT5G="${EXT_DIR}/cc.mqtt5.generated/install"
export BM="${PROOT}/bm_config.cmake"
export BROOT="/home/ubuntu/boost_1_85_0"
cmake .. -DCMAKE_BUILD_TYPE=Release \
    -DCMAKE_PREFIX_PATH="${COMMS}\;${MQTT5G}" \
    -DCC_MQTT5_CUSTOM_CLIENT_CONFIG_FILES="${BM}" \
    -DBOOST_ROOT= "${BROOT}" \
    -DCC_MQTT5_CLIENT_LIB_FORCE_PIC=ON
CMake Warning:
  Ignoring extra path from command line:

   "/home/ubuntu/boost_1_85_0"


-- The C compiler identification is GNU 13.2.0
-- The CXX compiler identification is GNU 13.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Error at CMakeLists.txt:33 (find_package):
  By not providing "FindLibComms.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "LibComms",
  but CMake did not find one.

  Could not find a package configuration file provided by "LibComms" with any
  of the following names:

    LibCommsConfig.cmake
    libcomms-config.cmake

  Add the installation prefix of "LibComms" to CMAKE_PREFIX_PATH or set
  "LibComms_DIR" to a directory containing one of the above files.  If
  "LibComms" provides a separate development package or SDK, be sure it has
  been installed.


-- Configuring incomplete, errors occurred!
```

Let me know if I can provide further information.

**Comment by arobenko** [comment]
The [CMAKE_PREFIX_PATH](https://cmake.org/cmake/help/v3.28/variable/CMAKE_PREFIX_PATH.html) configuration variable is expected to point to the **installation** directory(ies) of the externals, not their source directories. Also if you try to compile the applications, there should be no space after the **-DBOOST_ROOT=**.

Try to use [full_debug_build_gcc.sh](https://github.com/commschamp/cc.mqtt5.libs/blob/master/script/full_debug_build_gcc.sh) script to test the build. If it's successful in your environment, try to analyze what it does and customize it to your needs.
```
cd /src/of/cc.mqtt5.libs
./script/full_debug_build.gcc.sh
```

**Comment by schkovich** [comment]
Thanks a bunch!

How did I get here? I run `BUILD_DIR=/home/ubuntu/cc.mqtt5.libs ./prepare_externals.sh` creating the following structure:
```sh
tree -d ~/cc.mqtt5.libs/externals/
/home/ubuntu/cc.mqtt5.libs/externals/
├── cc.mqtt5.generated
│   ├── build
│   │   ├── CMakeFiles
│   │   │   ├── 3.28.3
│   │   │   │   ├── CompilerIdC
│   │   │   │   │   └── tmp
│   │   │   │   └── CompilerIdCXX
│   │   │   │       └── tmp
│   │   │   ├── CMakeScratch
│   │   │   ├── Export
│   │   │   │   └── 0d59e8a6f0c5601755b0e21b80938926
│   │   │   └── pkgRedirects
│   │   └── install
│   │       ├── include
│   │       │   └── cc_mqtt5
│   │       │       ├── dispatch
│   │       │       ├── factory
│   │       │       ├── field
│   │       │       ├── frame
│   │       │       │   └── layer
│   │       │       ├── input
│   │       │       ├── message
│   │       │       └── options
│   │       └── lib
│   │           └── cc_mqtt5
│   │               └── cmake
│   ├── doc
│   └── include
│       └── cc_mqtt5
│           ├── dispatch
│           ├── factory
│           ├── field
│           ├── frame
│           │   └── layer
│           ├── input
│           ├── message
│           └── options
└── comms
    ├── build
    │   ├── CMakeFiles
    │   │   ├── 3.28.3
    │   │   │   ├── CompilerIdC
    │   │   │   │   └── tmp
    │   │   │   └── CompilerIdCXX
    │   │   │       └── tmp
    │   │   ├── CMakeScratch
    │   │   ├── Export
    │   │   │   └── c6c6a5e993b75fa418c832763a342ede
    │   │   └── pkgRedirects
    │   ├── install
    │   │   ├── include
    │   │   │   └── comms
    │   │   │       ├── details
    │   │   │       ├── field
    │   │   │       │   ├── adapter
    │   │   │       │   ├── basic
    │   │   │       │   └── details
    │   │   │       ├── protocol
    │   │   │       │   ├── checksum
    │   │   │       │   └── details
    │   │   │       └── util
    │   │   │           └── details
    │   │   └── lib
    │   │       └── LibComms
    │   │           └── cmake
    │   └── test
    │       └── CMakeFiles
    ├── cmake
    ├── doc
    ├── doxygen
    │   └── dia
    ├── include
    │   └── comms
    │       ├── details
    │       ├── field
    │       │   ├── adapter
    │       │   ├── basic
    │       │   └── details
    │       ├── protocol
    │       │   ├── checksum
    │       │   └── details
    │       └── util
    │           └── details
    ├── script
    └── test
```
I applied suggested changes but it didn't work.
```sh
export PROOT="/home/ubuntu/cc.mqtt5.libs"
export EXT_DIR="${PROOT}/externals"
export COMMS="${EXT_DIR}/comms/build/install"
export MQTT5G="${EXT_DIR}/cc.mqtt5.generated/build/install"
export BM="${PROOT}/bm_config.cmake"
export BROOT="/home/ubuntu/boost_1_85_0"
cmake .. -DCMAKE_BUILD_TYPE=Release \
    -DCMAKE_PREFIX_PATH="${COMMS}\;${MQTT5G}" \
    -DCC_MQTT5_CUSTOM_CLIENT_CONFIG_FILES="${BM}" \
    -DBOOST_ROOT="${BROOT}" \
    -DCC_MQTT5_CLIENT_LIB_FORCE_PIC=ON
```
Nevertheless, your comment made me wonder, what the difference is between my setup and the one in the pipeline. I mimicked the setup from the pipeline.
```sh
EXTERNALS_DIR="/home/ubuntu/.local/share/cc.mqtt5/externals" \
COMMON_INSTALL_DIR="/home/ubuntu/.local/share/cc.mqtt5/build/install" \
BUILD_DIR="/home/ubuntu/.local/share/cc.mqtt5/build" \
./prepare_externals.sh
```
I find the resulting directory structure more clear. The source and built artefacts are separated.
```sh
~/.local/share$ tree -d cc.mqtt5/
cc.mqtt5/
├── build
│   └── install
│       ├── include
│       │   ├── cc_mqtt5
│       │   │   ├── dispatch
│       │   │   ├── factory
│       │   │   ├── field
│       │   │   ├── frame
│       │   │   │   └── layer
│       │   │   ├── input
│       │   │   ├── message
│       │   │   └── options
│       │   └── comms
│       │       ├── details
│       │       ├── field
│       │       │   ├── adapter
│       │       │   ├── basic
│       │       │   └── details
│       │       ├── protocol
│       │       │   ├── checksum
│       │       │   └── details
│       │       └── util
│       │           └── details
│       └── lib
│           ├── LibComms
│           │   └── cmake
│           └── cc_mqtt5
│               └── cmake
└── externals
    ├── cc.mqtt5.generated
    │   └── build
    │       └── CMakeFiles
    │           ├── 3.28.3
    │           │   ├── CompilerIdC
    │           │   │   └── tmp
    │           │   └── CompilerIdCXX
    │           │       └── tmp
    │           ├── CMakeScratch
    │           ├── Export
    │           │   └── 0d59e8a6f0c5601755b0e21b80938926
    │           └── pkgRedirects
    └── comms
        └── build
            ├── CMakeFiles
            │   ├── 3.28.3
            │   │   ├── CompilerIdC
            │   │   │   └── tmp
            │   │   └── CompilerIdCXX
            │   │       └── tmp
            │   ├── CMakeScratch
            │   ├── Export
            │   │   └── c6c6a5e993b75fa418c832763a342ede
            │   └── pkgRedirects
            └── test
                └── CMakeFiles
```
I removed the custom Boost install to reduce the complexity.
```sh
export PROOT="/home/ubuntu/.local/share/cc.mqtt5"
export EXT_DIR="${PROOT}/externals"
export PREFIX_PATH="${PROOT}/build/install"
export BM="${PROOT}/bm_config.cmake"
export BROOT="/home/ubuntu/boost_1_85_0"
cmake .. -DCMAKE_BUILD_TYPE=Release \
    -DCMAKE_PREFIX_PATH="${PREFIX_PATH}" \
    -DCC_MQTT5_CUSTOM_CLIENT_CONFIG_FILES="${BM}" \
    -DBOOST_ROOT="${BROOT}" \
    -DCC_MQTT5_CLIENT_LIB_FORCE_PIC=ON
```
That worked as expected. 

The only significant difference is that in the first case, a [list](https://cmake.org/cmake/help/latest/manual/cmake-language.7.html#lists) of paths is provided in `CMAKE_PREFIX_PATH`. I did not have time to try it, but the escaped semicolon might be causing the problem. According to the documentation 

> The sequence \\; does not divide a value but is replaced by ; in the resulting element.

I did not run `cmake --build . --config Release --target install` since that would install the library on the local system, and that's not what I need. I am considering creating a [package](https://cmake.org/cmake/help/v3.28/manual/cmake-packages.7.html#id8). Do I have a better and easier option to use the built library on an embedded system?

**Comment by arobenko** [comment]
All the CommsChampion Ecosystem libraries are expected to create their own packages when built. It means when the right path(s) is set in the **CMAKE_PREFIX_PATH** the following packages are expected to be found:

- `find_package(LibComms)` - finds [COMMS Library](https://github.com/commschamp/comms) and creates `cc::comms` cmake target to link against.
- `find_package(cc_mqtt5)` - finds [cc.mqtt5.generated](https://github.com/commschamp/cc.mqtt5.generated) MQTT v5 protocol definition library and craetes `cc::cc_mqtt5` cmake target to link against

The above are dependencies for this project and when this project is built, it creates another cmake configuration file allowing to do:
- `find_package(cc_mqtt5_client)` - it creates `cc::cc_mqtt5_client` cmake target to link against.

It can be used in your own project to link against the provided MQTT5 client library.

Based on your comment I guess your knowledge of cmake is someone lacking at this stage. You can pass [CMAKE_INSTALL_PREFIX](https://cmake.org/cmake/help/v3.28/variable/CMAKE_INSTALL_PREFIX.html) variable to your cmake configuration to set a custom install directory without worrying that the installation will corrupt your host system. That's what I do in my build scripts.

I also recommend taking a closer look at my [cc.cmake](https://github.com/commschamp/cc.cmake) project. It can be used to compile the required MQTT v5 client library without worrying of its dependencies, which will be built automatically. Just enable it in your configuration file and cmake will take care of the rest:
```
set (OPT_CC_MQTT5_LIBS ON)
list (APPEND CC_MQTT5_LIBS_CMAKE_ARGS "-DCC_MQTT5_CLIENT_LIB_FORCE_PIC=ON")
list (APPEND CC_MQTT5_LIBS_CMAKE_ARGS "-DCC_MQTT5_CUSTOM_CLIENT_CONFIG_FILES=/path/to/bare/metal/config")
...
``` 
The [cc.cmake](https://github.com/commschamp/cc.cmake) project can be pre-built using your bare-metal toolchain or used as an [external project](https://cmake.org/cmake/help/latest/module/ExternalProject.html) inside your own cmake configuration (you will need to customly define your `cc::cc_mqtt5_client` cmake target because it won't be ready yet)

**Comment by schkovich** [comment]
> Based on your comment I guess your knowledge of cmake is someone lacking at this stage.

I'm afraid you are overestimating my knowledge of cmake. ;) I've been using it as a shell command for a long, but that's it.

We did the PoC of an IoT project using the Arduino framework and cloud platform, but that's not a sellable product. We have to get out of the Arduino Cloud at least. Our infrastructure is in CloudFlare. Cloudflare is running a MQTT v5 broker. There is no v5 MQTT client for the Arduino framework. Probably MQTT v3 would work, but with no V5 features. As an old stubborn bastard, I decided to bite the bullet. I will get it eventually. ;)

Thank you for helping.
